sw/device/lib/crypto/impl/rsa/rsa_keygen.c

```diff
@@ -0,0 +1,157 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/crypto/impl/rsa/rsa_keygen.h"
+
+#include "sw/device/lib/base/hardened.h"
+#include "sw/device/lib/base/hardened_memory.h"
+#include "sw/device/lib/crypto/drivers/otbn.h"
+
+// Module ID for status codes.
+#define MODULE_ID MAKE_MODULE_ID('r', 'k', 'g')
+
+OTBN_DECLARE_APP_SYMBOLS(run_rsa_keygen);         // The OTBN RSA keygen binary.
+OTBN_DECLARE_SYMBOL_ADDR(run_rsa_keygen, mode);   // Application mode.
+OTBN_DECLARE_SYMBOL_ADDR(run_rsa_keygen, rsa_n);  // Public exponent n.
+OTBN_DECLARE_SYMBOL_ADDR(run_rsa_keygen, rsa_d);  // Private exponent d.
+
+static const otbn_app_t kOtbnAppRsaKeygen = OTBN_APP_T_INIT(run_rsa_keygen);
+static const otbn_addr_t kOtbnVarRsaMode =
+    OTBN_ADDR_T_INIT(run_rsa_keygen, mode);
+static const otbn_addr_t kOtbnVarRsaN = OTBN_ADDR_T_INIT(run_rsa_keygen, rsa_n);
+static const otbn_addr_t kOtbnVarRsaD = OTBN_ADDR_T_INIT(run_rsa_keygen, rsa_d);
+
+enum {
+  /* Fixed public exponent for generated keys. This exponent is 2^16 + 1, also
+     known as "F4" because it's the fourth Fermat number. */
+  kFixedPublicExponent = 65537,
+  /* Number of words used to represent the application mode. */
+  kOtbnRsaModeWords = 1,
+};
+
+// Available application modes. Must match the values from `run_rsa_keygen.s`.
+// TODO: I think it's possible to pull these values directly from symbols in
+// the binary. See:
+//   hw/ip/otbn/util/otbn_objdump.py -t
+//   bazel-out/k8-fastbuild-ST-2cc462681f62/bin/sw/otbn/crypto/run_rsa_keygen.rv32embed.o
+static const uint32_t kOtbnRsaMode2048 = 0x3b7;
+static const uint32_t kOtbnRsaMode3072 = 0x4fa;
+static const uint32_t kOtbnRsaMode4096 = 0x74d;
+
+status_t rsa_keygen_2048_start(void) {
+  // Load the RSA key generation app. Fails if OTBN is non-idle.
+  HARDENED_TRY(otbn_load_app(kOtbnAppRsaKeygen));
+
+  // Set mode to RSA-2048.
+  HARDENED_TRY(
+      otbn_dmem_write(kOtbnRsaModeWords, &kOtbnRsaMode2048, kOtbnVarRsaMode));
+
+  // Start the OTBN routine.
+  return otbn_execute();
+}
+
+status_t rsa_keygen_2048_finalize(rsa_2048_public_key_t *public_key,
+                                  rsa_2048_private_key_t *private_key) {
+  // Spin here waiting for OTBN to complete.
+  HARDENED_TRY(otbn_busy_wait_for_done());
+
+  // Read the public modulus (n) from OTBN dmem.
+  HARDENED_TRY(
+      otbn_dmem_read(kRsa2048NumWords, kOtbnVarRsaN, private_key->n.data));
```

Comment:
It would probably make sense just to harden `otbn_dmem_read` in all cases; it shouldn't ever be performance-critical. Created https://github.com/lowRISC/opentitan/issues/18333 to track.